framework/src/transfers/MultiAppGeneralFieldNearestNodeTransfer.C

```diff
@@ -58,97 +63,118 @@ MultiAppGeneralFieldNearestNodeTransfer::prepareEvaluationOfInterpValues(
-    auto & from_mesh = from_problem.mesh(_displaced_source_mesh);
-    MooseVariableFieldBase & from_var =
-        from_problem.getVariable(0,
-                                 _from_var_names[var_index],
-                                 Moose::VarKindType::VAR_ANY,
-                                 Moose::VarFieldType::VAR_FIELD_ANY);
-
-    System & from_sys = from_var.sys().system();
-    unsigned int from_var_num = from_sys.variable_number(getFromVarName(var_index));
-    // FEM type info
-    auto & fe_type = from_sys.variable_type(from_var.number());
-    bool is_nodal = fe_type.family == LAGRANGE;
-
-    if (is_nodal)
+    // Nest a loop on apps only for the Positions case. Regular case only looks at one app
+    unsigned int num_apps_to_consider = _nearest_positions_obj ? _from_problems.size() : 1;
+    for (unsigned int app_index = 0; app_index < num_apps_to_consider; app_index++)
     {
-      for (const auto & node : from_mesh.getMesh().local_node_ptr_range())
+      unsigned int i_from;
+      if (!_nearest_positions_obj)
+        i_from = i_source;
+      else
+        i_from = app_index;
+      // NOTE: If we decide we do not want to mix apps when using the nearest-positions
+      // we could continue here when the positions is not the nearest to the app
+
+      FEProblemBase & from_problem = *_from_problems[i_from];
+      auto & from_mesh = from_problem.mesh(_displaced_source_mesh);
+      MooseVariableFieldBase & from_var =
+          from_problem.getVariable(0,
+                                   _from_var_names[var_index],
+                                   Moose::VarKindType::VAR_ANY,
+                                   Moose::VarFieldType::VAR_FIELD_ANY);
+
+      System & from_sys = from_var.sys().system();
+      unsigned int from_var_num = from_sys.variable_number(getFromVarName(var_index));
+      // FEM type info
+      auto & fe_type = from_sys.variable_type(from_var.number());
+      bool is_nodal = fe_type.family == LAGRANGE;
+
+      if (is_nodal)
       {
-        if (node->n_dofs(from_sys.number(), from_var_num) < 1)
-          continue;
+        for (const auto & node : from_mesh.getMesh().local_node_ptr_range())
+        {
+          if (node->n_dofs(from_sys.number(), from_var_num) < 1)
+            continue;
 
-        if (!_from_blocks.empty() && !inBlocks(_from_blocks, from_mesh, node))
-          continue;
+          if (!_from_blocks.empty() && !inBlocks(_from_blocks, from_mesh, node))
+            continue;
 
-        if (!_from_boundaries.empty() && !onBoundaries(_from_boundaries, from_mesh, node))
-          continue;
+          if (!_from_boundaries.empty() && !onBoundaries(_from_boundaries, from_mesh, node))
+            continue;
 
-        points[i_from].push_back(*node + _from_positions[i_from]);
-        auto dof = node->dof_number(from_sys.number(), from_var_num, 0);
+          if (_nearest_positions_obj &&
+              !closestToPosition(i_source, *node + _from_positions[i_from]))
```

Comment:
Pretty cool!

Reply:
From the conversation thread in the input file, which re-confused me, I'd like a comment here about what this is doing conceptually. Here is how I currently understand this:

Take two 1D apps from a MultiApp, which both span along the x-axis. Let's say app 1 is to the right of app 0. We draw a perpendicular cut line equidistant between the "positions" of the apps. What I understand this condition here as doing is disqualifying any nodes from app 0 that fall the right of the cut line, and any nodes from app 1 that fall to the left of the cut line. Is my interpretation correct?

Reply:
No it should not be the positions of the app. It s the positions from the positions object. 

But you're right that this is used to disqualify nodes. 
The area closer to a given position than any other position in the object will be used to group (partition?) source points. Then when you look for a value for a target point, you find the position it is the closest to, and use those corresponding source points

I ll add comments in the code on top of doco

Reply:
> No it should not be the positions of the app. It s the positions from the positions object.

Yea I understand that the multiapp and the position are different objects and the distinction is important. But I guess that when I think of the positions system I think that there is an app in the multiapp object that is associated with a position  in the positions object. Is that fair?

Reply:
(yet here I am reviewing this ... 🤔 )

Reply:
> Is that fair?

no it s not fair. That would have been nice.
Think of April's HTGR : one app doing heat transfer, 400k pebbles so 400k positions. 
And we need to be distinguishing these "nearest pebble" areas to make sure we dont send data from one pebble to another

Reply:
however your thinking is correct for ShapeEvaluation and UO transfer. In those cases, the source of the data is the entire app, not a group of nodes. So the positions are essentially matched to one app.

But there's a need there too, for example for the 400k apps 400k pebble case. Positions let us use the nearest app to gather data from.

Reply:
Ok sorry I meant to indicate a one way relationship. When a multiapp is using the `positions_objects` parameter, then an app in the multiapp will be associated with a position from (one of) the positions object. Is that right?

> Think of April's HTGR : one app doing heat transfer, 400k pebbles so 400k positions.

I don't think I understand this yet. Ok you have one app modeling heat transfer physics for 400k pebbles. And so I can imagine a `Positions` in that input file for the 400k pebbles. But what data transfer is that `Positions` object controlling? Data exchange within that single input? Or with another multiapp?

Reply:
> Positions let us use the nearest app to gather data from.

Yea I get this. And that's very cool.

> a group of nodes

This is still what I need to understand. I should look at the tests you introduced in your last PR. I imagine that added a test in which a position is logically associated with a group of nodes (or elements) as opposed to an app in a multiapp

Reply:
so filling your kd trees you are looping over your `num_sources` in the outer loop and then the inner loop is over nodes so O(n_sources * n_nodes). What you'd ideally want is to loop over nodes and then do a KD tree search to figure out the closest position match as opposed to what seems to currently be a linear search over the positions. E.g. you want to do kd tree searches to fill your kd trees 😆 It seems like this kind of algorithm improvement would be really beneficial for when you go to something like 400k pebbles/400k positions

Reply:
I can see KDTrees being created by Positions objects and used for getNearestPosition

however, at this stage, I think it s early for such an optimization.

Wrt to the ordering of the loop, if we do a loop on nodes, then we need to cache all the information on the entire node loop then create the KDTrees. It makes sense but there s a memory cost to that

Reply:
> then we need to cache all the information on the entire node loop then create the KDTrees.

I'm not sure I understand this, but yea I'm fine with not optimizing at this point

Reply:
So you can’t form KD trees as you are looping on the nodes afaik. You have to take all the data relevant to each KDtree and use it all at once to create the tree. 

So if we loop on nodes rather than trees., we need to cache the datat for all the trees as we are looping on nodes